sw/host/opentitanlib/src/util/openocd.rs

```diff
@@ -0,0 +1,110 @@
+use std::io::Write;
+use std::process::{Command, Stdio, Child};
+use crate::io::jtag::{Jtag, JtagParams};
+
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use std::cell::Cell;
+
+use std::net::TcpStream;
+
+use crate::impl_serializable_error;
+
+use std::{thread, time};
+
+
+/// Represents an OpenOCD server and we can interact with interactively
+pub struct OpenOcdClient {
+    /// OpenOCD child process
+    child: Child,
+    /// TcpStream to the telnet port of OpenOCD
+    telnet_stream: Cell<Option<TcpStream>>,
+}
+
+/// Errors related to the OpenOCD client.
+#[derive(Error, Debug, Deserialize, Serialize)]
+pub enum OpenOcdClientError {
+    #[error("No OpenOCD binary path provided")]
+    OpenOcdBinaryRequired,
+    #[error("Generic error {0}")]
+    Generic(String),
+}
+impl_serializable_error!(OpenOcdClientError);
+
+impl OpenOcdClient {
+    pub fn new(opts: &JtagParams) -> Result<OpenOcdClient> {
+        let mut args = Vec::new();
+        if let Some(cfg) = &opts.openocd_adapter_config {
+            args.push("-f");
+            args.push(&cfg);
+        }
+        // Tell OpenOCD how to connect to the lifecycle controller's RISC-V TAP.
+        // This works because lc_ctrl uses an instance of the RISC-V debug
+        // module [0]. Its IRLEN is hardcoded to 5 [1] and its IDCODE is 1 [2].
+        //
+        // [0]: https://docs.opentitan.org/hw/ip/lc_ctrl/doc/#life-cycle-tap-controller
+        // [1]: See where `dmi_jtag_tap` is created in hw/vendor/pulp_riscv_dbg/src/dmi_jtag.sv
+        // [2]: See where `dmi_jtag` is instantiated in hw/ip/lc_ctrl/rtl/lc_ctrl.sv
+        //
+        // Because these commands are command-line args, they *must* end with
+        // semicolons to satisfy the TCL lexer. When a TCL program is read from
+        // a file, newlines are sufficient to separate statements.
+        args.extend(vec![
+            "-c",
+            "transport select jtag;",
+            "-c",
+            "jtag newtap lc_ctrl tap -irlen 5 -expected-id 0x00000001 -ignore-bypass;",
+            "-c",
+            "target create lc_ctrl.tap.0 riscv -chain-position lc_ctrl.tap -rtos hwthread;",
+            "-c",
+            "init;",
+            "-c",
+            "scan_chain;"
+        ]);
+        log::info!("CWD: {:?}", std::env::current_dir());
+        log::info!(
+            "Spawning OpenOCD: {:?} {:?}",
+            opts.openocd,
+            args.join(" ")
+        );
+        let child = Command::new(&opts.openocd)
+            .args(args)
+            .stdin(Stdio::null())
+            .stdout(Stdio::inherit())
+            .stderr(Stdio::inherit())
+            .spawn().context("failed to spawn openocd")?;
+        thread::sleep(time::Duration::from_millis(1000));
+        Ok(OpenOcdClient {
+            child,
+            telnet_stream: None.into(),
+        })
+    }
+
+    fn connect_telnet(&self) -> Result<()> {
+        log::info!("Connecting to OpenOCD telnet interface...");
```

Comment:
I would think this would automatically happen when the OpenOCD server is started no? I.e., we would have `server_start()` method that would both start the OpenOCD server process and connect otlib to it, preparing the system to start accepting commands.

Reply:
Yes.